core/domain/auth_domain.py

```diff
@@ -45,28 +46,147 @@ class AuthClaims(python_utils.OBJECT):
+        self.parent_user_id = parent_user_id
+        self.deleted = deleted
+
+    @property
+    def auth_id(self):
+        """Returns the platform-agnostic authentication identifier of a user."""
+        return self.firebase_auth_id or self.gae_id
+
+    def validate(self):
+        """Checks that user_id, gae_id, firebase_auth_id, and parent_user_id
+        fields of this UserAuthDetails domain object are valid.
+
+        Raises:
+            ValidationError. The user_id is not str.
+            ValidationError. The gae_id is not str.
+            ValidationError. The firebase_auth_id is not str.
+            ValidationError. The parent_user_id is not str.
+        """
+        if not isinstance(self.user_id, python_utils.BASESTRING):
+            raise utils.ValidationError(
+                'Expected user_id to be a string, received %s' % self.user_id)
+        if not self.user_id:
+            raise utils.ValidationError('No user id specified.')
+        if not utils.is_user_id_valid(self.user_id):
+            raise utils.ValidationError('user_id=%r is in a wrong format.' % (
+                self.user_id))
+
+        if (self.gae_id is not None and
+                not isinstance(self.gae_id, python_utils.BASESTRING)):
+            raise utils.ValidationError(
+                'Expected gae_id to be a string, received %s' % self.gae_id)
+
+        if (self.firebase_auth_id is not None and
+                not isinstance(self.firebase_auth_id, python_utils.BASESTRING)):
+            raise utils.ValidationError(
+                'Expected firebase_auth_id to be a string, received %s' %
+                self.firebase_auth_id)
+
+        if (self.parent_user_id is not None and
+                not utils.is_user_id_valid(self.parent_user_id)):
+            raise utils.ValidationError(
+                'The parent user ID is in a wrong format.')
+
+        if self.parent_user_id and self.auth_id:
+            raise utils.ValidationError(
+                'The parent user ID and auth_id cannot be present together '
+                'for a user.')
+
+        if not self.parent_user_id and not self.auth_id:
+            raise utils.ValidationError(
+                'The parent user ID and auth_id cannot be None together '
+                'for a user.')
+
+    def is_full_user(self):
+        """Whether the user is a full user (not a profile user).
+
+        Returns:
+            bool. True if user is full user, False otherwise.
+        """
+        return self.auth_id is not None
+
+    def to_dict(self):
+        """Returns a dict matching the properties of UserAuthDetailsModel."""
+        return {
+            'gae_id': self.gae_id,
+            'firebase_auth_id': self.firebase_auth_id,
+            'parent_user_id': self.parent_user_id,
+            'deleted': self.deleted
+        }
+
+    @classmethod
+    def from_model(cls, user_auth_details_model):
+        """Returns new domain object with values from a UserAuthDetailsModel."""
```

Comment:
Add Args, Returns.

Also is this a general thing we do already? I didn't think we had finalized the structure here yet and I don't want this to be the only class which does it. If this is one-off please do it the usual way in the corresponding services file for now. /cc @vojtechjelinek